cleanup created docker containers

mikz opened the pull request:
Cancelled or timeouted workflow would keep the docker container running. Closes game-ci/unity-test-runner#197

This has two parts:

Part one. The entrypoints.

`runs.post`: GitHub Action metadata allow running something after the action (regardless of a failure, crash, timeout, ...). https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runspost

However, it needs to be a `.js` file and it can't be configured to pass any arguments.

There already was `index.js` used as the main entrypoint. The build process of this file uses typescript compiler and ncc to pack all dependencies into one .js file. And ncc has no way of generating multiple files in one go, so the only solution would be to run ncc twice and generate two independent files.

That would be quite unfortunate, wasting time and storage. So I rather came up with a new entrypoint that symlinked from two locations. And this new entrypoint understands how it was executed, so it can run the correct behaviour. This makes it easy to add `runs.pre` if needed.

This new entrypoint is in `index.ts`. The original `index.ts` is now in `main.ts`.

Part two.
The signals. I've tried:
* try/catch/finally around the `await Docker.run`. Catch and finally are not executed when process receives SIGINT. See the discussion in: https://github.com/nodejs/node/discussions/29480
* New AbortController and AbortSignal. Great concept, but the action.exec does not support it. So it can't be aborted.
* Doing cleanup on `process.on('exit')`. Unfortunately you can't really do async stuff from there, so can't really run the docker rm command to delete the container.
* Using `process.on('SIGINT')`. For some reason that wasn't really executing for me. I'd not put my hand in fire for this, but I assume because it was in the signal handler it does something special, or would heed to be scheduled for later with `setTimeout(0)`.

Evaluating all these I came to a conclusion that it is fragile and just relying on a `runs.post` is much better and safer. 


I've not tested this on Windows, so that probably needs to happen. I've tried running locally with `act` and it worked. Worked also on my hosted linux runner.

webbertakken commented:
Great explanation! Thanks for the elaborate details, it will be very helpful for merging this.

> Evaluating all these I came to a conclusion that it is fragile and just relying on a runs.post is much better and safer.

Ok, based on your research I would agree. Let's go for your approach.

> Using process.on('SIGINT'). For some reason that wasn't really executing for me. I'd not put my hand in fire for this, but I assume because it was in the signal handler it does something special, or would heed to be scheduled for later with setTimeout(0).

As I understand it, that's probably because an event listener doesn't stop the event from bubbling and you only have a few milliseconds to do any cleanup.

webbertakken commented on src/index.ts:
```diff
@@ -1,60 +1,21 @@
-    const baseImage = new ImageTag({ editorVersion, customImage });
-    const runnerTemporaryPath = process.env.RUNNER_TEMP;
+async function run([_, name]: string[]) {
+  const script = basename(name ?? 'main.js');
```
prefer a separate line instead of using null coalesce operator (which can be hard to understand for new developers, especially when having to consider all possible execution paths)

mikz replied:
Sure.

mikz replied:
Addressed.

webbertakken commented on src/model/docker.ts:
```diff
@@ -1,8 +1,24 @@
 import { exec } from '@actions/exec';
 import path from 'path';
 
+const cidfilePath = parameters => {
```
would you mind renaming this to `containerIdFilePath` please?

GabLeRoux replied:
Personally, I'm ok with the current name since docker's own flag is `--cidfile`, but I agree it makes it easier to understand what the concept is.

mikz replied:
I'm happy to write a few comments to document why it is doing what it is doing, but personally I'd rather keep it `cidfile` so it is searchable. 
I've started with `containerIDPath`, but in the end it read a bit wrong to refer to the same concept with multiple names.

webbertakken replied:
If you just change this method into `containerIdFilePath`, the rest of the code will also be clear. And you wouldn't need comments, because your code is self descriptive. Note that we don't use ID in both caps but only the first letter of acronyms as caps (camelCase).

mikz replied:
Addressed.

webbertakken commented on src/post.ts:
```diff
@@ -0,0 +1,16 @@
+    const runnerTemporaryPath = process.env.RUNNER_TEMP;
+    const githubAction = process.env.GITHUB_ACTION;
+
+    await Docker.cleanup({
+      runnerTemporaryPath,
+      githubAction,
+    });
```
I was thinking perhaps we could call this method something with `ensure...` in the name. What do you think?

mikz replied:
I think `Docker.removeContainer` would be quite explanatory. WDYT?

webbertakken replied:
It makes it look like removal always happens in this bit, which isn't the case. I think ensure makes sense here.

mikz replied:
Addressed.

webbertakken commented on dist/main.js:
```diff
@@ -0,0 +1 @@
+index.js
```
How does this work exactly? Is this executed as a bash script or is this some kind of magic javascript or actions syntax?

mikz replied:
This is how GitHub shows symlinks. 

Both `main.js` and `post.js` are symlinked to `index.js`. 

So when you run `node main.js` it will technically evaluate the `index.js`, but it can read with what arguments it was evaluated with and act accordingly. 

Github action runs `node $FULL_PATH/main.js`, so the `index.js` will take the first argument and see what file was executed and run it.

mikz replied:
Addressed by a comment.

webbertakken reviewed (approved):
Looks pretty solid to me. Awesome work!

GabLeRoux commented on dist/post.js:
```diff
@@ -0,0 +1 @@
+index.js
```
Same question as https://github.com/game-ci/unity-test-runner/pull/198/files#r1012351548

GabLeRoux commented on src/post.ts:
```diff
@@ -0,0 +1,16 @@
+
+export async function run() {
+  try {
+    const runnerTemporaryPath = process.env.RUNNER_TEMP;
```
I asked myself what's `RUNNER_TEMP`, it's documented here:  
https://docs.github.com/en/actions/learn-github-actions/environment-variables

> RUNNER_TEMP | The path to a  temporary directory on the runner. This directory is emptied at the  beginning and end of each job. Note that files will not be removed if  the runner's user account does not have permission to delete them. For  example, D:\a\_temp
> -- | --

mikz replied:
I've taken this from the already existing code that relies on it too. 

Ideally this would be extracted into own function, so it is evaluated the same on all the branches. 
But I didn't want to rewrite the whole docker.ts, since I didn't find any unit tests and wanted to minimize the risk of breaking something.

webbertakken replied:
You can use https://github.com/game-ci/unity-test-runner/blob/main/src/model/action.ts to `get tempPath`

mikz replied:
Ok, I've actually implemented something that ended up in Input, but it clearly belongs to the Action. Going to move it there.

mikz replied:
I did not want to refactor the command generation of docker which should be changed to use this new tempPath.

It also deserves to not be text, but rather just an array, so it can be passed to the action/exec as args and it would correctly do all the escaping of arguments.

Also it would be easy to reuse some common base for the linux and win commands (cidfile, env, ...).

mikz commented:
I've updated it to clear up things brought in the review. 

I've also added some type annotations to the new code, so it explicitly handles default values.
IMO that is important in case this ever gets executed outside of Github Action Runner (like unit tests, or trying it locally).

webbertakken commented:
LGTM. This can be merged once it's tested.

mikz commented:
I'm running this on my hosted runner with concurrency limited cancellable workflows and it works well.
Don't have Windows pipeline to try it on.
